bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/EphemerisUpdater.scala

```diff
@@ -68,63 +68,122 @@ object EphemerisUpdater {
+          hd  <- nst.horizonsDesignation.toList
+        } yield (hd, spt, nst)
 
-        val updates: HS2[HS2[Unit]] =
-          nstArgs.zipWithIndex
-                 .traverseU { case ((spt, nst, hd), n) => update(spt, nst, hd, n) }
-                 .map(_.sequenceU.void)
+        UpdateContext(obsN, site, tocN, toc, nsts)
+      }
+    }
 
-        updates.map(_ *> HS2.delay(tocN.setDataObject(toc)))
+  }
 
-      case None => HS2.unit.point[HS2]
+  // Computes an HS2 action that looks up ephemerides for all nonsidereal
+  // targets in the list of observations, displaying status in the glass pane UI.
+  // Returns a map that can be used after the fact to lookup the `Ephemeris`
+  // associated with a horizons key and site.
+  private def bulkLookup(
+    obsNs: List[ISPObservation],
+    start: Long,
+    ui:    UI
+  ): HS2[Map[(HorizonsDesignation, Site), Ephemeris]] = {
+
+    val keys = obsNs.flatMap(o => UpdateContext.fromObservation(o).toList).flatMap { uc =>
+      uc.nsts.map { case (hd, _, _) => (hd, uc.site) }
+    }.distinct
+
+    val total = keys.size
+
+    keys.zipWithIndex.traverseU { case ((hd, site), n) =>
+      for {
+        _ <- ui.show(s"Looking up ephemeris ${n+1} of $total ...")
+        e <- lookup(hd, site, start)
+      } yield ((hd, site), e)
+    }.map(_.toMap)
 
-    }
+  }
 
   /**
-   * Fetch updated ephemerides for all nonsidereal targets in `obs`, showing status in the root pane
-   * associated with the given `Component`, returning *another* program that completes the update
-   * without any further network access (i.e., quickly) and clears out the UI.
+   * Fetches updated ephemerides for all nonsidereal targets in the list of
+   * observations, showing status in the root pane associated with the given
+   * `Component`, returning a function that given a single observation computes
+   * *another* action that completes the update without any further network
+   * access (i.e., quickly).  The idea is that we can perform the final update
+   * actions quickly while holding the program lock.
    */
-  def refreshEphemerides(obsN: ISPObservation, start: Long, c: Component): IO[IO[Unit]] = {
+  def refreshAll(
+    obsNs:  List[ISPObservation],
+    start:  Long,
+    c:      Component
+  ): IO[ISPObservation => IO[Unit]] = {
+
+    def setEphemerides(
+      obsN: ISPObservation,
+      m:    Map[(HorizonsDesignation, Site), Ephemeris]
+    ): IO[Unit] =
+      UpdateContext.fromObservation(obsN).fold(IO.ioUnit) { uc =>
+        uc.updateAction(m).withResultLogging(Log).run.map(_.leftMap(_ => ()).merge)
+      }
+
     val ui = UI(c)
-    refreshEphemerides(obsN, start, ui).withResultLogging(Log).run.map {
-      case -\/(e) => IO.ioUnit
-      case \/-(a) => (a.withResultLogging(Log) ensuring ui.hide).run.void
+    (bulkLookup(obsNs, start, ui).withResultLogging(Log) ensuring ui.hide).run.map {
+      case -\/(_) => (obsN: ISPObservation) => IO.ioUnit
+      case \/-(m) => setEphemerides(_, m)
```

Comment:
Here the behavior differs a bit from before in that the glass pane is dismissed after the lengthy horizons lookup but before the actual observation nodes are updated.  The observation node updates are quick though so I deemed this to be an unimportant detail.

Reply:
I should say that since usually they'll be setting the scheduling block for multiple observations of the *same* nonsidereal target i thought it was important not to make them pay the price of waiting for the ephemeris lookup multiple times.  So the `bulkLookup` extracts all the nonsidereal target (horizons id, site) pairs and does the lookup just once for each one.